File: chat/src/main/java/org/thoughtcrime/securesms/providers/MyBlobProvider.java
package org.thoughtcrime.securesms.providers;

import android.content.Context;
import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;

import com.difft.android.base.log.lumberjack.L;
import com.difft.android.base.utils.FileUtil;

import util.StreamUtil;
import util.concurrent.TTExecutors;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Objects;
import java.util.UUID;
import java.util.concurrent.Future;

public class MyBlobProvider {
    private static final String TAG = L.INSTANCE.tag(MyBlobProvider.class);

    private static final MyBlobProvider INSTANCE = new MyBlobProvider();

    public static MyBlobProvider getInstance() {
        return INSTANCE;
    }

    public BlobBuilder forData(@NonNull byte[] data) {
        return new BlobBuilder(new ByteArrayInputStream(data));
    }

    public BlobBuilder forData(@NonNull InputStream data) {
        return new BlobBuilder(data);
    }

    public synchronized @NonNull InputStream getStream(@NonNull Context context, @NonNull Uri uri) throws IOException {
        File file = new File(uri.getPath());
        if (!file.exists()) {
            throw new IOException("File does not exist for URI: " + uri);
        }
        return new FileInputStream(file);
    }

    public void delete(@NonNull Uri uri) {
        File file = new File(Objects.requireNonNull(uri.getPath()));
        if (file.exists()) {
            file.delete();
        }
    }

    public class BlobBuilder {

        private InputStream data;
        private String id;
        private String mimeType;
        private String fileName;

        private BlobBuilder(@NonNull InputStream data) {
            this.id = UUID.randomUUID().toString();
            this.data = data;
        }

        public BlobBuilder withMimeType(@NonNull String mimeType) {
            this.mimeType = mimeType;
            return this;
        }

        public BlobBuilder withFileName(@Nullable String fileName) {
            this.fileName = fileName;
            return this;
        }

        @WorkerThread
        public Future<Uri> createForDraftAttachmentAsync(@NonNull Context context)
                throws IOException {
            File outputFile = new File(FileUtil.INSTANCE.getFilePath(FileUtil.DRAFT_ATTACHMENTS_DIRECTORY), buildFileName(id));
            OutputStream outputStream = new FileOutputStream(outputFile);

            final Uri uri = buildUri(context);

            return TTExecutors.BOUNDED.submit(() -> {
                try {
                    StreamUtil.copy(data, outputStream);
                    return uri;
                } catch (IOException e) {
                    delete(context, uri);
                    L.w(e, () -> "Error during write!");
                    throw e;
                }
            });
        }


        private @NonNull String buildFileName(@NonNull String id) {
            String suffix = "";
            if (mimeType != null) {
                switch (mimeType) {
                    case "image/jpeg":
                        suffix = ".jpg";
                        break;
                    case "image/png":
                        suffix = ".png";
                        break;
                    case "video/mp4":
                        suffix = ".mp4";
                        break;
                    case "audio/aac":
                        suffix = ".aac";
                        break;
                    default:
                        suffix = ".blob";
                        break;
                }
            }
            if (fileName != null) {
                return fileName;
            } else {
                return id + suffix;
            }
        }

        private @NonNull Uri buildUri(@NonNull Context context) {
            return Uri.fromFile(new File(FileUtil.INSTANCE.getFilePath(FileUtil.DRAFT_ATTACHMENTS_DIRECTORY), buildFileName(id)));
        }

//        private File getOrCreateDirectory(@NonNull Context context, @NonNull String directory) {
//            return new File(FileUtil.INSTANCE.getFilePath(FileUtil.FILE_DIR_ATTACHMENT));
//        }

        private void delete(@NonNull Context context, @NonNull Uri uri) {
            File file = new File(uri.getPath());
            if (file.exists()) {
                file.delete();
            }
        }
    }
}

File: chat/src/main/java/com/difft/android/chat/providers/MyBlobProvider.kt
package com.difft.android.chat.providers

import android.content.ContentResolver
import android.content.Context
import android.net.Uri
import androidx.annotation.WorkerThread
import com.difft.android.base.concurrent.AppExecutors
import com.difft.android.base.log.lumberjack.L
import com.difft.android.base.utils.AppPrivateStorage
import com.difft.android.base.utils.FileUtil
import util.StreamUtil
import java.io.ByteArrayInputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.util.UUID
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionException
import java.util.concurrent.Future

class MyBlobProvider private constructor() {

    fun forData(data: ByteArray): BlobBuilder = BlobBuilder(ByteArrayInputStream(data))

    fun forData(data: InputStream): BlobBuilder = BlobBuilder(data)

    @Synchronized
    @Throws(IOException::class)
    fun getStream(context: Context, uri: Uri): InputStream {
        val file = File(uri.path!!)
        if (!file.exists()) {
            throw IOException("File does not exist for URI: $uri")
        }
        return FileInputStream(file)
    }

    /**
     * Deletes a blob owned by this provider. Ownership is structural: a file-scheme uri whose path
     * resolves inside an app-private root. A uri outside that domain is REFUSED rather than
     * attempted, so shared-storage media stays untouched by construction instead of by a delete
     * that merely happens to fail — and whose failure nobody could observe.
     *
     * The gate lives here and nowhere else: callers pass the uri they own, this is the only place
     * that decides whether the path belongs to us.
     *
     * No `exists()` pre-check: `File.delete()` already reports false for a missing file, so the
     * probe only bought an extra IO round trip while letting existence take part in the decision.
     *
     * @return true only when a file was actually removed.
     */
    fun delete(uri: Uri): Boolean {
        val scheme = uri.scheme
        // A null scheme is legal input: callers build uris straight from bare absolute paths. Only
        // schemes that clearly belong to someone else are refused, which also keeps a content uri
        // out of File(): its path is a pseudo path such as /external/images/media/123.
        if (scheme != null && scheme != ContentResolver.SCHEME_FILE) {
            L.w { "[Blob] delete refused: unsupported scheme=$scheme" }
            return false
        }
        val path = uri.path
        if (path.isNullOrBlank()) {
            L.w { "[Blob] delete refused: uri has no path" }
            return false
        }
        if (!AppPrivateStorage.isAppPrivate(path)) {
            // Deliberately without the path: a shared-storage path carries the user's own file name.
            L.w { "[Blob] delete refused: path is outside app-private storage" }
            return false
        }
        val deleted = try {
            File(path).delete()
        } catch (e: SecurityException) {
            L.w { "[Blob] delete threw: ${e.javaClass.simpleName}" }
            false
        }
        if (!deleted) {
            L.w { "[Blob] delete returned false (already gone or denied)" }
        }
        return deleted
    }

    inner class BlobBuilder internal constructor(private val data: InputStream) {

        private val id: String = UUID.randomUUID().toString()
        private var mimeType: String? = null
        private var fileName: String? = null

        fun withMimeType(mimeType: String): BlobBuilder {
            this.mimeType = mimeType
            return this
        }

        fun withFileName(fileName: String?): BlobBuilder {
            this.fileName = fileName
            return this
        }

        @WorkerThread
        @Throws(IOException::class)
        fun createForDraftAttachmentAsync(context: Context): Future<Uri> {
            val outputFile = File(FileUtil.getFilePath(FileUtil.DRAFT_ATTACHMENTS_DIRECTORY), buildFileName(id))
            val outputStream: OutputStream = FileOutputStream(outputFile)
            val uri = buildUri(context)

            return CompletableFuture.supplyAsync({
                try {
                    StreamUtil.copy(data, outputStream)
                    uri
                } catch (e: IOException) {
                    // Delegates to the outer delete so the write-failure cleanup path gets the same
                    // ownership gate and the same failure log as every other caller.
                    delete(uri)
                    L.w(e) { "Error during write!" }
                    // Wrap so Future.get() callers still see the original IOException
                    // via ExecutionException.getCause().
                    throw CompletionException(e)
                }
            }, AppExecutors.Default)
        }

        private fun buildFileName(id: String): String {
            var suffix = ""
            if (mimeType != null) {
                suffix = when (mimeType) {
                    "image/jpeg" -> ".jpg"
                    "image/png" -> ".png"
                    "video/mp4" -> ".mp4"
                    "audio/aac" -> ".aac"
                    else -> ".blob"
                }
            }
            return fileName ?: (id + suffix)
        }

        private fun buildUri(context: Context): Uri =
            Uri.fromFile(File(FileUtil.getFilePath(FileUtil.DRAFT_ATTACHMENTS_DIRECTORY), buildFileName(id)))
    }

    companion object {
        private val INSTANCE = MyBlobProvider()

        @JvmStatic
        fun getInstance(): MyBlobProvider = INSTANCE
    }
}
